io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -442,16 +441,154 @@ impl Lvol {
+            &mut attr_descrs,
+            snap_param.clone(),
+            &mut cstrs,
+        )?;
+
+        let c_snapshot_name = snap_param.name().unwrap().into_cstring();
+
+        // No need to flush blob's buffers explicitly as SPDK always
+        // synchronizes blob when taking a snapshot.
+        unsafe {
+            vbdev_lvol_create_snapshot_ext(
+                self.as_inner_ptr(),
+                c_snapshot_name.as_ptr(),
+                attr_descrs.as_mut_ptr(),
+                SnapshotXattrs::COUNT as u32,
+                Some(done_cb),
+                done_cb_arg,
+            )
+        };
+        Ok(())
+    }
+    /// Common API to get the xattr from blob.
+    pub fn get_blob_xattr(
+        snapshot: &Lvol,
+        attr: &SnapshotXattrs,
+    ) -> (String, bool) {
+        let mut valid_snapshot = true;
+        let mut val = std::ptr::null();
+        let mut size: u64 = 0;
+        let attr_id = attr.name().to_string().into_cstring();
+        let curr_attr_val;
+        unsafe {
+            let blob = snapshot.bs_iter_first();
+            let r = spdk_blob_get_xattr_value(
+                blob,
+                attr_id.as_ptr(),
+                &mut val as *mut *const c_void,
+                &mut size as *mut u64,
+            );
+            // mark snapshot invalid, if any attribute not found
+            if r != 0 {
+                valid_snapshot = false;
+                warn!(
+                    "Snapshot attribute {:?} not found, snapshot{:?}",
+                    attr_id, snapshot
+                );
+                return (String::default(), valid_snapshot);
+            }
+            curr_attr_val = String::from_raw_parts(
+                val as *mut u8,
+                size as usize,
+                size as usize,
+            );
         }
+        (curr_attr_val, valid_snapshot)
+    }
+    /// List All Snapshot.
+    pub fn list_all_snapshots() -> Vec<VolumeSnapshotDescriptor> {
+        let mut snapshot_list: Vec<VolumeSnapshotDescriptor> = Vec::new();
+        let bdev =
+            UntypedBdev::bdev_first().expect("Failed to enumerate devices");
+
+        let lvol_devices = bdev
+            .into_iter()
+            .filter(|b| b.driver() == "lvol")
+            .map(|b| Lvol::try_from(b).unwrap())
+            .collect::<Vec<Lvol>>();
+        // return empty snapshot parameter list, if no snapshot found.
+        if lvol_devices.len() <= 1 {
+            return snapshot_list;
+        }
+        for lvol in lvol_devices {
+            // skip lvol if it is not snapshot.
+            if !lvol.is_snapshot() {
+                continue;
+            }
+            let mut valid_snapshot = true;
+            let mut snapshot_param: SnapshotParams = Default::default();
+            for attr in SnapshotXattrs::iter() {
+                let curr_attr_val;
+                (curr_attr_val, valid_snapshot) =
+                    Self::get_blob_xattr(&lvol, &attr);
```

Comment:
hmm valid_snapshot keeps getting overridden on each iteration of the for loop right?
Should it be "sticky" ?
Otherwise, rather than having it returned from get_xattr this could we simply have a fn in VolumeSnapshotDescriptor that checks if any param is none or empty and expose a self.valid() ?

Reply:
Accepted. Changed the logic.